### ISSUE TITLE
Query Parser has difficulty parsing columns that contain json or arrays of json

### ISSUE THREAD

**Issue by sarahzinger** [grafana/azure-data-explorer-datasource#353]
**What happened**:
We are showing users invalid Where filters

**What you expected to happen**:
I was expecting all dropdown options in the query editor lead to valid syntax

**How to reproduce it (as minimally and precisely as possible)**:
- have a column that contains json (in our internal test db of DestinyActivities look at the Teams column). Here is an example:
```
| id | Teams |

| --------------- | --------------- |

| 123 | {"18":{"Standing":"Victory","Score":5,"TeamID":0},"19":{"TeamID":0,"Standing":"Defeat","Score":3}} |
```
Our Query builder tries to parse this json with Where filters like so:
![Screen Shot 2022-04-11 at 8 57 19 PM](https://user-images.githubusercontent.com/6620164/162857267-dfaaf790-24e0-4f10-9e84-6f8b6eedfdcf.png)
- however if you try any of these filters, you'll hit a syntax error. You'll see that the KQL query we send off is:
```
DestinyActivities
| where $__timeFilter(TimeGenerated)
| where Teams.18.Score == 0
| order by TimeGenerated asc
```
when I think it should actually be something like this
```
DestinyActivities
| where $__timeFilter(TimeGenerated)
| where todynamic(Teams)["18"]["Score"] == 0
| order by TimeGenerated asc
```

You can see the same thing happen for columns with arrays of json:
- have a column that contains an array of json (in our internal test data at grafana look at the Players column in DestinyActivities) which is formated like so:
```
| id | Players |

| --------------- | --------------- |

| 123 | [{"Values":{"timePlayedSeconds":97,"completed":0,"completionReason":255,"efficiency":0},"Score":0}] |
```
- in our query builder dropdown you'll see in a Where filter that we have some odd options available to choose from:
![Screen Shot 2022-04-11 at 8 01 20 PM](https://user-images.githubusercontent.com/6620164/162852201-bca6b35d-7e40-4865-9d90-71e6eb8524ab.png)
- if you try to make a query with any of these options, we get back a syntax error

A current workaround for this would be to let users edit the KQL query directly.

I think it's valid to say that our Query Builder can not handle every possible data format (what if you want to query an array of arrays of arrays of arrays of json of json of json, forever and ever, how would you even render that in a Where filter), I do think it's a bit odd to show users options that will always result in a syntax error as they are not valid column types. I wonder is it possible for us to:
- filter out options from our Where column that are not valid 
- add good documentation on what columns our query builder supports/doesn't support and make that documentation easily accessible within the app
- and then maybe add better support for arrays, json, and arrays of json

**Environment**:

- Grafana version: 8.5(ish)
- Plugin version: 3.7.1

TODO:

 - [x] Disable dynamic columns while we implement a working solution
 - [x] Add support for simple objects (objects with no arrays)
 ~~- [ ] Improve UX for objects. The UI should show nested groups not "flattened" strings.~~ Turns out that ReactSelect [do not support deeply nested groups](https://github.com/JedWatson/react-select/discussions/4315), which makes it difficult to create a solution in which an object is clickable so leaving this out for the moment.
 - [x] Add support for arrays.
 ~~- [ ] [TBD] Improve UX for arrays.~~

**Comment by aangelisc** [comment]
Looking into this a bit today hear are some thoughts:

- KQL can support querying columns of arbitrary complexity, but the more complex the columns to be queried on then the more complex the query
- Supporting columns that contain simple JSON objects (including nested objects but not objects where arrays are values of properties) should be quite straightforward as you've mentioned above (using `todynamic`)
- Supporting columns that contain arrays of simple types should be possible using something akin to `mv-apply`
- Supporting columns that contain arrays of simple JSON objects should also be possible as above

An example query for an array of JSON objects:
```
DestinyActivities
| where $__timeFilter(TimeGenerated)
| extend Players = Players
| mv-apply Players on (
    where todynamic(Players)["Values"]["efficiency"] == 0
)
```

The `mv-apply` function acts like a map, carrying out the subquery for each element of `Players`. We then parse the element as a JSON object and carry out the where clause. 

Thoughts on this @sarahzinger?

**Comment by andresmgot** [comment]
I think that we can address this on different iterations. The first step would be to detect if a field is a JSON object (regardless if it's an object or an array) and remove it from the suggestion list since it won't work anyway. We should also document the limitation.

Then we can take a look at how to work with those values. We should either support any possible JSON object or none since a partial support could lead to buggy behaviors. For that we can ask Microsoft and check if they have a good suggestion of how to do that with Kusto if needed.

**Comment by aangelisc** [comment]
That makes sense @andresmgot. Further investigation shows that this issue doesn't just impact the `Where` filters, but also the `Aggregate` and `Group by` functions as well:

![image](https://user-images.githubusercontent.com/15019026/165475504-4918f079-7228-4caf-8efe-2487e9f3ff5d.png)
![image](https://user-images.githubusercontent.com/15019026/165475612-b7351ef6-731e-47bd-8f3c-2554a90a849f.png)

**Comment by andresmgot** [comment]
Released in 4.1.0
